vnext/local-cli/src/runWindows/utils/msbuildtools.ts

```diff
@@ -135,60 +134,92 @@ class MSBuildTools {
       await fs.promises.unlink(errorLog);
     }
   }
-}
 
-function VSWhere(
-  requires: string,
-  version: string,
-  property: string,
-  verbose: boolean,
-): string {
-  // This path is maintained and VS has promised to keep it valid.
-  const vsWherePath = path.join(
-    process.env['ProgramFiles(x86)'] || process.env.ProgramFiles!,
-    '/Microsoft Visual Studio/Installer/vswhere.exe',
-  );
+  static findAvailableVersion(
+    buildArch: BuildArch,
+    verbose: boolean,
+  ): MSBuildTools {
+    // https://aka.ms/vs/workloads
+    const requires = [
+      'Microsoft.Component.MSBuild',
+      getVCToolsByArch(buildArch),
+    ];
+    const version = process.env.VisualStudioVersion || '16.0';
+    const vsInstallation = findLatestVsInstall({requires, version, verbose});
+
+    if (!vsInstallation) {
+      if (process.env.VisualStudioVersion != null) {
+        throw new Error(
+          `MSBuild tools not found for version ${
+            process.env.VisualStudioVersion
+          } (from environment). Make sure all required components have been installed`,
+        );
+      } else {
+        throw new Error(
+          'MSBuild tools not found. Make sure all required components have been installed',
+        );
+      }
+    }
 
-  if (verbose) {
-    console.log('Looking for vswhere at: ' + vsWherePath);
-  }
+    const toolsPath = path.join(
+      vsInstallation.installationPath,
+      'MSBuild/Current/Bin',
+    );
 
-  // Check if vswhere is present and try to find MSBuild.
-  if (fs.existsSync(vsWherePath)) {
-    if (verbose) {
-      console.log('Found vswhere.');
+    if (fs.existsSync(toolsPath)) {
+      newSuccess(
+        `Found MSBuild v${version} at ${toolsPath} (${
+          vsInstallation.installationVersion
+        })`,
+      );
+      return new MSBuildTools(
+        version,
+        toolsPath,
+        vsInstallation.installationVersion,
+      );
+    } else {
+      throw new Error(`MSBuild path '${toolsPath} does not exist'`);
     }
-    const propertyValue = child_process
-      .execSync(
-        `"${vsWherePath}" -version [${version},${Number(version) +
-          1}) -products * -requires ${requires} -property ${property}`,
-      )
-      .toString()
-      .split(EOL)[0];
-    return propertyValue;
-  } else {
-    if (verbose) {
-      console.log("Couldn't find vswhere, querying registry.");
+  }
+
+  static getAllAvailableUAPVersions(): Version[] {
```

Comment:
This is the same as before, just moved upward in the file to mark it as part of the class.